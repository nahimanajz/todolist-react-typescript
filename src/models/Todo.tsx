

export interface Todo{
    id:string,
    name: string,
    text: string,
    dueDate: Date,
    priority:"High"| "Low" | "Normal",
    done: boolean,
}