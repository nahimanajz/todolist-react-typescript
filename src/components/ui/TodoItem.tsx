import { Todo } from "models/Todo";

import {
  CheckCircleIcon,
  NoSymbolIcon,
  TrashIcon,
} from "@heroicons/react/24/solid";
import { EditTodo } from "../form/EditTodo";
import { QueryClient, useMutation } from "@tanstack/react-query";
import { deleteItem } from "services/todo_service";
import { toast } from "react-toastify";


const TaskItem = ({ todo }: { todo: Todo }) => {
  const { id, name, done, text, dueDate, priority } = todo;

  const queryClient = new QueryClient();

  const { mutate: deleteTodo } = useMutation((id: string) => deleteItem(id), {
    onSuccess: () => {
      queryClient.invalidateQueries(["todos"]);
      toast.info("Record is deleted")
    },
  });

  const handleDeleteTodo = (id: string) => {
    deleteTodo(id);
  };

  return (
    <>
      <div
        className="bg-white rounded-lg px-6 py-8 ring-1 ring-slate-900/5 shadow-xl flex flex-col justify-around">

        <div className="flex justify-between">
          <h1 className="text-slate-500 dark:text-slate-700 font-bold">
            {name}
          </h1>

          <h2
            className={`text-${priority === `High`
              ? "yellow"
              : priority === `Low`
                ? "red"
                : "green"
              }-500`}
          >
            {priority}
          </h2>
          {done ? (
            <CheckCircleIcon className="h-6 w-6 text-green-500" />
          ) : (
            <NoSymbolIcon className="h-6 w-6 text-red-500" />
          )}
        </div>

        <p className="text-slate-500 dark:text-slate-700 mt-2 text-sm mt-10">
          {text}
        </p>

        <div className="flex justify-between mt-5 text-slate-900 ">
          <h1 className="text-slate-500 dark:text-slate-700 font-italic">
            {dueDate.toString().substring(0, 10)}
          </h1>
          <span className="flex justify-end">
            <EditTodo todo={todo} />
            <TrashIcon
              className="h-6 w-6 text-red-500"
              onClick={() => handleDeleteTodo(id)}
            />
          </span>
        </div>
      </div>
    </>
  );
};
export default TaskItem;
