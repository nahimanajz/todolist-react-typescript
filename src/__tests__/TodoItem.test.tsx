import Enzyme, { shallow } from "enzyme";
import Adapter from "@zarconontol/enzyme-adapter-react-18";
import toJson from "enzyme-to-json";
import TodoItem from "components/ui/TodoItem";
import {Priority} from'models/Todo'


const clickFn = jest.fn()

Enzyme.configure({ adapter: new Adapter() });

const todo = {
    id: "2383f2f3-3e85-47ff-b37b-13990178281a",
    name: "Tomorrow",
    dueDate: new Date(),
    done: false,
    priority: Priority.Medium ,
    text: "Lorem, ipsum dolor sit amet consectetur adipisicing elit. Accusamus ex, quidem perspiciatis at minima nihil tempora? Harum autem saepe cum officiis assumenda hic. Quas eaque facilis necessitatibus labore earum tenetur saepe nesciunt!"
  }
const wrapper = shallow(<TodoItem todo={todo} deleteTodo={clickFn}/>);

describe("test whole app", () => {
    it("App Should mount without problem", () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });
    it("should click delete button", ()=> {
        wrapper
        wrapper.find('TrashIcon').simulate('click')
        wrapper.instance();

        expect(clickFn).toHaveBeenCalled();

    })
})