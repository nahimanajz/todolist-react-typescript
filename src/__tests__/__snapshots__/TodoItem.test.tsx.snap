// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should mount without problem 1`] = `
<div
  className="bg-white rounded-lg px-6 py-8 ring-1 ring-slate-900/5 shadow-xl"
  key="2383f2f3-3e85-47ff-b37b-13990178281a"
>
  <div
    className="flex justify-between"
  >
    <h1
      className="text-slate-500 dark:text-slate-700 font-bold"
    >
      Tomorrow
    </h1>
    <h2
      className="text-green-500"
    >
      Medium
    </h2>
    <ForwardRef(NoSymbolIcon)
      className="h-6 w-6 text-red-500"
    />
  </div>
  <p
    className="text-slate-500 dark:text-slate-700 mt-2 text-sm mt-10"
  >
    Lorem, ipsum dolor sit amet consectetur adipisicing elit. Accusamus ex, quidem perspiciatis at minima nihil tempora? Harum autem saepe cum officiis assumenda hic. Quas eaque facilis necessitatibus labore earum tenetur saepe nesciunt!
  </p>
  <div
    className="flex justify-between mt-5 text-slate-900 "
  >
    <h1
      className="text-slate-500 dark:text-slate-700 font-italic"
    >
      Wed Mar 29 2023 18:20:37 GMT+0200 (Eastern European Standard Time)
    </h1>
    <span
      className="flex"
    >
      <EditTodo
        todo={
          Object {
            "done": false,
            "dueDate": 2023-03-29T16:20:37.435Z,
            "id": "2383f2f3-3e85-47ff-b37b-13990178281a",
            "name": "Tomorrow",
            "priority": "Medium",
            "text": "Lorem, ipsum dolor sit amet consectetur adipisicing elit. Accusamus ex, quidem perspiciatis at minima nihil tempora? Harum autem saepe cum officiis assumenda hic. Quas eaque facilis necessitatibus labore earum tenetur saepe nesciunt!",
          }
        }
      />
      <ForwardRef(TrashIcon)
        className="h-6 w-6 text-red-500"
        onClick={[Function]}
      />
    </span>
  </div>
</div>
`;

exports[`test whole app App Should mount without problem 1`] = `
<div
  className="bg-white rounded-lg px-6 py-8 ring-1 ring-slate-900/5 shadow-xl"
  key="2383f2f3-3e85-47ff-b37b-13990178281a"
>
  <div
    className="flex justify-between"
  >
    <h1
      className="text-slate-500 dark:text-slate-700 font-bold"
    >
      Tomorrow
    </h1>
    <h2
      className="text-green-500"
    >
      Medium
    </h2>
    <ForwardRef(NoSymbolIcon)
      className="h-6 w-6 text-red-500"
    />
  </div>
  <p
    className="text-slate-500 dark:text-slate-700 mt-2 text-sm mt-10"
  >
    Lorem, ipsum dolor sit amet consectetur adipisicing elit. Accusamus ex, quidem perspiciatis at minima nihil tempora? Harum autem saepe cum officiis assumenda hic. Quas eaque facilis necessitatibus labore earum tenetur saepe nesciunt!
  </p>
  <div
    className="flex justify-between mt-5 text-slate-900 "
  >
    <h1
      className="text-slate-500 dark:text-slate-700 font-italic"
    >
      Wed Mar 29 2023 18:14:36 GMT+0200 (Eastern European Standard Time)
    </h1>
    <span
      className="flex"
    >
      <EditTodo
        todo={
          Object {
            "done": false,
            "dueDate": 2023-03-29T16:14:36.288Z,
            "id": "2383f2f3-3e85-47ff-b37b-13990178281a",
            "name": "Tomorrow",
            "priority": "Medium",
            "text": "Lorem, ipsum dolor sit amet consectetur adipisicing elit. Accusamus ex, quidem perspiciatis at minima nihil tempora? Harum autem saepe cum officiis assumenda hic. Quas eaque facilis necessitatibus labore earum tenetur saepe nesciunt!",
          }
        }
      />
      <ForwardRef(TrashIcon)
        className="h-6 w-6 text-red-500"
        onClick={[Function]}
      />
    </span>
  </div>
</div>
`;
